core/test/utils/assertions.hpp

```diff
@@ -999,6 +999,38 @@ ::testing::AssertionResult array_equal(const std::string& first_expression,
 }
 
 
+template <typename ValueType>
+::testing::AssertionResult segmented_array_equal(
+    const std::string& first_expression, const std::string& second_expression,
+    const segmented_array<ValueType>& first,
+    const segmented_array<ValueType>& second)
+{
+    auto view_first =
+        gko::make_const_array_view(first.get_executor(), first.get_size(),
+                                   first.get_const_flat_data())
+            .copy_to_array();
+    auto view_second =
+        gko::make_const_array_view(second.get_executor(), second.get_size(),
+                                   second.get_const_flat_data())
+            .copy_to_array();
+
+    auto buffer_result = array_equal(first_expression, second_expression,
+                                     view_first, view_second);
+    if (buffer_result == ::testing::AssertionFailure()) {
+        return buffer_result << "Buffers of the segmented arrays mismatch";
+    }
+
+    auto offsets_result =
+        array_equal(first_expression, second_expression, first.get_offsets(),
+                    second.get_offsets());
+    if (offsets_result == ::testing::AssertionFailure()) {
+        return offsets_result << "Offsets of the segmented arrays mismatch";
+    }
```

Comment:
I would probably check offsets first, since if they mismatch, the rest of the contents make even less sense.